exonum/src/api_ng/node/private/mod.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Private part of the Exonum rest api.
+
+use std::{collections::HashMap, net::SocketAddr};
+
+use api_ng::{Error as ApiError, ServiceApiScope, ServiceApiState};
+use blockchain::{Service, SharedNodeState};
+use crypto::PublicKey;
+use messages::PROTOCOL_MAJOR_VERSION;
+use node::ExternalMessage;
+
+#[doc(hidden)]
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct ServiceInfo {
+    name: String,
+    id: u16,
+}
+
+/// `DTO` is used to transfer information about node.
+#[doc(hidden)]
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct NodeInfo {
+    pub core_version: Option<String>,
+    pub protocol_version: u8,
+    pub services: Vec<ServiceInfo>,
+}
+
+impl NodeInfo {
+    /// Creates new `NodeInfo`, from services list.
+    pub fn new<'a, I>(services: I) -> NodeInfo
+    where
+        I: IntoIterator<Item = &'a Box<Service>>,
+    {
+        let core_version = option_env!("CARGO_PKG_VERSION").map(|ver| ver.to_owned());
+        NodeInfo {
+            core_version,
+            protocol_version: PROTOCOL_MAJOR_VERSION,
+            services: services
```

Comment:
I'd suggest to move logic out of object creating. Why not collect services at first and then just pass it to the `NodeInfo`?

Reply:
I think it's pretty common pattern in Rust. Don't have personal opinion though